### DISCUSSION TITLE
[QUESTION] Detect connection

### DISCUSSION THREAD

**Discussion by hernanr99** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1733] [Q&A]
Hi, I'm trying to detect when the other end loses the connection, what I'm seeing is that the library takes a long time to detect that connection change (it takes around 30 seconds). Could you give me a way to detect that connection change more quickly?

**Comment by hmfarimani** [comment]
Can you use ping command via ip?

**Comment by disa6302** [comment]
@hernanr99 ,

Can you be specific? Are you talking about the signaling connection?

**Reply by tplusy** [reply]
I have this issue too, #1998 RTC_PEER_CONNECTION_STATE_DISCONNECTED  comes delay more than 30s

**Reply by sirknightj** [reply]
Note that RTC_PEER_CONNECTION_STATE_DISCONNECTED would mean that the connection can no longer be used. If the PeerConnection hasn't switched to that state yet, it is still usable assuming the network conditions are restored to their original condtions (e.g. brief internet disconnect). You can perform an experiment by using this master sample with the JS test page and "force TURN". Disconnect your internet for a couple seconds then reconnect and you should see that the media exchange resumes.

One approach to quickly detect disconnects could be to use a heartbeat function over the data channel. The viewer can send a ping message every second or so, then have the master respond with a message in response. The master sample application by default echoes the incoming messages back to the viewer https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/4f1d3add4c0f88212fe6646898e5aba140ecbb3c/samples/Common.c#L1892.

Another possible approach is to poll the sender report statistics and seeing how many frames have come in since the last time you polled it https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/getStats and https://developer.mozilla.org/en-US/docs/Web/API/RTCVideoSourceStats.
